src/main/java/com/danielagapov/spawn/Services/OAuth/OAuthService.java

```
@@ -17,39 +19,47 @@ public class OAuthService implements IOAuthService {
     private final IUserIdExternalIdMapRepository externalIdMapRepository;
     private final IUserService userService;
 
+
     public OAuthService(IUserRepository userRepository, IUserIdExternalIdMapRepository externalIdMapRepository, IUserService userService) {
         this.userRepository = userRepository;
         this.externalIdMapRepository = externalIdMapRepository;
         this.userService = userService;
     }
 
-    @Override
-    public UserDTO verifyUser(OAuth2User oauthUser) {
-        UserDTO user = unpackOAuthUser(oauthUser);
-
-        String externId = oauthUser.getAttribute("sub");
-        if (externId == null) throw new ApplicationException("Subject was null");
+    public AbstractUserDTO verifyUser(OAuth2User oauthUser) {
+        TempUserDTO tempUser = unpackOAuthUser(oauthUser);
+        UserIdExternalIdMap mapping = getMapping(tempUser);
 
-        UserIdExternalIdMap mapping = externalIdMapRepository.findById(externId).orElse(null);
-        if (mapping == null) {
-            User userEntity = UserMapper.toEntity(user);
-            User savedEntity = userRepository.save(userEntity);
+        return mapping == null ? tempUser : getUserDTO(mapping);
+    }
 
-            UserIdExternalIdMap newMapping = new UserIdExternalIdMap(externId, savedEntity);
-            externalIdMapRepository.save(newMapping);
+    // REQUIRES: userDTO.id() == externId from unpackOAuthUser()
```

Could you throw an exception instead of having this "requires"? Better to bullet-proof, than rely on us using the method properly, in the CPSC 210 style, in my opinion

---

In that method, we don't have access to the external id because we aren't persisting the TempUser so I can't make that check for whether to throw an exception. To make this work, mobile should be sending back the same id as is being received in the response to calling /google/sign-in endpoint. It should be almost guaranteed for the requires clause to be true, as long as Mobile doesn't generate their own user ids before sending it to the OAuthController.

---

Ok, thanks for clarifying. We can leave it be